bust-a-move-framework/src/main/java/nl/tudelft/bust_a_move20162017/bust_a_move_framework/arena/Arena.java

```diff
@@ -29,542 +29,540 @@
+
+        // loop over each bubble to see if newVisited contains it
+        // if it does not, remove the bubble
+        for (int i = 0; i < bubble2DArray.size(); i++) {
+            Bubble[] row = bubble2DArray.get(i);
+            for (Bubble bubble : row) {
+                if (!newVisited.contains(bubble)) {
+                    removeBubble(bubble);
+                    dropped_bubbles++;
+                }
+            }
+        }
+        if (dropped_bubbles > 0) {
+            App.game.player.score.scoreBubblesDropped(dropped_bubbles);
+        }
+
+        return visited;
+        //return null;
+    }
+
+    /**
+     * Checks for bubbles to drop using a recursive algorithm
+     *
+     * @param visited
+     * @return
+     */
+    private LinkedList<Bubble> checkBubblesToDrop(LinkedList<Bubble> visited) {
+        boolean addedSomething = false;
+        for (int i = 0; i < visited.size(); i++) {
+            Bubble b = visited.get(i);
+            if (b == null) continue;
+            Bubble[] neighbours = getNeighbors(getRow(b.getY()), getColumn(b.getX(), b.getY()));
+            for (Bubble nb : neighbours) {
+                if (!visited.contains(nb)) {
+                    visited.addLast(nb);
+                    addedSomething = true;
+                }
+            }
+        }
+
+        if (!addedSomething) return visited;
+
+        // else do a recursive call with the new list
+        visited = checkBubblesToDrop(visited);
+
+        return visited;
+
+
+    }
+
+    /**
+     * Retrieves the neighbors of the Bubble
+     *
+     * @param row    row-th entry of the linkedlist
+     * @param column column-th entry of the array
+     * @return array with the neighbor bubbles. If a neighbor does not exist, a null pointer is stored.
+     */
+    private Bubble[] getNeighbors(int row, int column) {
+        Bubble[] neighbors = new Bubble[6];
+
+        int height = bubble2DArray.size();
+        if (height <= row) {
+            return neighbors;
+        }
+        int width = bubble2DArray.get(row).length;
+
+        neighbors[0] = (column != 0) ? (bubble2DArray.get(row)[column - 1]) : null;
+        neighbors[3] = (column < width - 1) ? (bubble2DArray.get(row)[column + 1]) : null;
+
+        try {
+            if (width == 8) {
+                neighbors[1] = (row != 0 && column != 0) ? (bubble2DArray.get(row - 1)[column - 1]) : null;
+                neighbors[2] = (row != 0 && column < width - 1) ? (bubble2DArray.get(row - 1)[column]) : null;
+                neighbors[4] = (row < height - 1 && column < width - 1) ? (bubble2DArray.get(row + 1)[column]) : null;
+                neighbors[5] = (row < height - 1 && column != 0) ? (bubble2DArray.get(row + 1)[column - 1]) : null;
+            } else {
+                neighbors[1] = (row != 0 && column < WIDTH_BUBBLES) ? (bubble2DArray.get(row - 1)[column]) : null;
+                neighbors[2] = (row != 0) ? (bubble2DArray.get(row - 1)[column + 1]) : null;
```

Comment:
CORRECTION: what if column = 6? (Because width can be only be max 7)

Just add -> && (column < width - 1)